src/Iterators.jl

```diff
@@ -894,4 +895,94 @@ macro chain(ex)
     Expr(:let, Expr(:block, cycleex...), states...)
 end
 
+
+type MergedIter
+    it1   # iterator 1
+    it2   # iterator 2
+    lt    # custom `<` operator to compare elements of 2 iterators    
+end
+
+MergeIter(it1, it2) = MergeIter(it1, it2, isless)
+
+type MergedState
+    hd1   # current head of 1st iterator
+    hd2   # current head of 2nd iterator
+    s1    # current state of 1st iterator
+    s2    # current state of 2nd iterator
+end
+
+
+function start(it::MergedIter)
+    s1 = start(it.it1)
+    s2 = start(it.it2)
+    # doing 1 step ahead and caching next element of each iter
+    hd1, s1 = safenext(it.it1, s1)
+    hd2, s2 = safenext(it.it2, s2)
+    return MergedState(hd1, hd2, s1, s2)
+end
+
+"""
+Like `next`, but may be called on iterator that is already done. In this case
+new item is `nothing` and new state is the same as previous. 
+"""
+function safenext(it, s)
+    if done(it, s)
```

Comment:
Isn't it enough to keep one element of each iterator in the state? next returns the smaller one and replenishes the state with a new element of that iterator (if not one of them is done(...)).

This is also very natural, compare

start{T}(r::UnitRange{T}) = oftype(r.start + one(T), r.start)
next{T}(r::UnitRange{T}, i) = (convert(T, i), i + one(T))
done{T}(r::UnitRange{T}, i) = i == oftype(i, r.stop) + one(T)